lib/Sema/SemaBounds.cpp

```diff
@@ -140,7 +140,56 @@ BoundsExpr *Sema::ConcretizeFromFunctionType(BoundsExpr *Expr,
   ExprResult ConcreteBounds = ConcretizeBoundsExpr(*this, Params).TransformExpr(Expr);
   if (ConcreteBounds.isInvalid()) {
     llvm_unreachable("unexpected failure in making bounds concrete");
-    Result = nullptr;
+    return nullptr;
+  }
+  else {
+    Result = dyn_cast<BoundsExpr>(ConcreteBounds.get());
+    assert(Result && "unexpected dyn_cast failure");
+    return Result;
+  }
+}
+
+namespace {
+  class ConcretizeBoundsExprWithArgs : public TreeTransform<ConcretizeBoundsExprWithArgs> {
+    typedef TreeTransform<ConcretizeBoundsExprWithArgs> BaseTransform;
+
+  private:
+    const ArrayRef<Expr *> Arguments;
+
+  public:
+    ConcretizeBoundsExprWithArgs(Sema &SemaRef, ArrayRef<Expr *> Args) :
+      BaseTransform(SemaRef),
+      Arguments(Args) { }
+
+    ExprResult TransformPositionalParameterExpr(PositionalParameterExpr *E) {
+      unsigned index = E->getIndex();
+      if (index < Arguments.size()) {
+        Expr *AE = Arguments[index];
+
+        // We may only substitute if this argument expression is
+        // a non-modifying expression.
+        if (!SemaRef.CheckIsNonModifyingExpr(AE,
+                                             Sema::NonModifiyingExprRequirement::NMER_Unknown,
+                                             /*ReportError=*/false))
+          return ExprResult();
+
+        return AE;
+      } else {
+        llvm_unreachable("out of range index for positional argument");
+        return ExprResult();
+      }
+    }
+  };
+}
+
+BoundsExpr *Sema::ConcretizeFromFunctionTypeWithArgs(BoundsExpr *Bounds, ArrayRef<Expr *> Args) {
+  if (!Bounds)
+    return Bounds;
+
+  BoundsExpr *Result;
+  ExprResult ConcreteBounds = ConcretizeBoundsExprWithArgs(*this, Args).TransformExpr(Bounds);
+  if (ConcreteBounds.isInvalid()) {
+    return nullptr;
```

Comment:
No `llvm_unreachable` here (unlike the other similar functions) because this will return nullptr if one of the arguments we need to substitute is a modifying expression.